mlflow/sklearn/utils.py

```diff
@@ -25,6 +32,16 @@ def _get_Xy(args, kwargs, X_var_name, y_var_name):
     return kwargs[X_var_name], kwargs[y_var_name]
 
 
+def _get_fit_Xy(trained_estimator, fit_args, fit_kwargs, fit_arg_names):
+    # In most cases, X_var_name and y_var_name become "X" and "y", respectively.
+    # However, certain sklearn models use different variable names for X and y.
+    X_var_name, y_var_name = fit_arg_names[:2]
+    X, y_true = _get_Xy(fit_args, fit_kwargs, X_var_name, y_var_name)
+    y_pred = trained_estimator.predict(X)
```

Comment:
Certain clustering models don't define `predict()`. For example: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html#sklearn.cluster.SpectralClustering.fit.

We should probably handle the case where `predict()` is undefined. Let's talk about this tomorrow. In conjunction with my other comment about clustering down below (where `y_true` may not be the correct input for most of these metrics functions), our best option may be to forgo logging clustering metrics for now and focus solely on regression / classification.

Reply:
Sounds reasonable!